// 55tec_integration_lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.flowHook = exports.listen = void 0;
require("reflect-metadata");
const request_1 = require("./model/protocol/integrator/request");
const resource_1 = require("./model/protocol/server/resource");
const action_1 = require("./model/metadata/action");
const operation_1 = require("./model/metadata/action/operation");
const model_1 = require("./model");
/**
 * @TODO add integration model definition
 */
async function listen(channel, handler) {
    channel.on('message', async (data) => {
        try {
            if (!request_1.isRequest(data))
                return;
            let result = await handler(data);
            return channel.send({
                id: data.id,
                response: {
                    code: '200',
                    body: result
                }
            });
        }
        catch (err) {
            return channel.send({
                id: data.id,
                response: {
                    code: err.code ? err.code : '500',
                    error: {
                        message: err.message,
                        name: err.name
                    }
                }
            });
        }
    });
}
exports.listen = listen;
async function flowHook(integrationClient, server, actions) {
    let prefix = '/flows';
    let router = new resource_1.Router();
    for (let action in actions) {
        let parser = actions[action];
        let [entity, operation] = action_1.parse(action);
        let handler = async (req, res) => {
            try {
                let resultSet = parser(req);
                for (let result of resultSet) {
                    let { credentials, data } = result;
                    let entities = data;
                    if (!(entities instanceof Array))
                        entities = [entities];
                    await Promise.all(entities.map(async (entity) => integrationClient.send(model_1.Integration.FLOW, result.action || action, {
                        credentials,
                        data: entity,
                    })));
                }
                return res.status(204).end();
            }
            catch (err) {
                return res.status(err.code || 500).json({ error: err.message });
            }
        };
        let method = 'post';
        if ([operation_1.Operation.CREATE, operation_1.Operation.UPDATE, operation_1.Operation.DELETE].includes(operation)) {
            switch (operation) {
                case operation_1.Operation.UPDATE:
                    method = 'put';
                    break;
                case operation_1.Operation.DELETE:
                    method = 'delete';
                    break;
            }
            router[method](`${prefix}/${entity}`, handler);
        }
        router.post(`${prefix}/${action}`, handler);
    }
    server.use(router);
}
exports.flowHook = flowHook;
