// 55tec_integration_lib/model/metadata/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetafieldType = exports.TransformType = exports.ConstraintType = exports.Lang = void 0;
var Lang;
(function (Lang) {
    Lang["PT_BR"] = "pt-BR";
})(Lang = exports.Lang || (exports.Lang = {}));
var ConstraintType;
(function (ConstraintType) {
    ConstraintType["IN"] = "in";
    ConstraintType["MIN"] = "min";
    ConstraintType["MAX"] = "max";
    ConstraintType["PATTERN"] = "pattern";
    ConstraintType["HAS"] = "has";
})(ConstraintType = exports.ConstraintType || (exports.ConstraintType = {}));
var TransformType;
(function (TransformType) {
    TransformType["MASK"] = "mask";
})(TransformType = exports.TransformType || (exports.TransformType = {}));
var MetafieldType;
(function (MetafieldType) {
    MetafieldType["TEXT"] = "text";
    MetafieldType["LIST"] = "list";
    MetafieldType["OBJECT"] = "object";
    MetafieldType["NUMBER"] = "number";
    MetafieldType["DATE"] = "date";
    MetafieldType["ID"] = "id";
    MetafieldType["BOOLEAN"] = "boolean";
    MetafieldType["REFERENCE"] = "reference";
})(MetafieldType = exports.MetafieldType || (exports.MetafieldType = {}));
