// 55tec_integration_lib/model/protocol/integrator/request/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRequest = exports.RequestError = exports.getNextCursor = exports.getCursorQuery = void 0;
function getCursorQuery(identifier, sort) {
    const fields = Object.keys(sort).sort();
    if (fields.length > 1)
        throw new Error('Cursors with multiple sort fields are not supported yet');
    if (!identifier)
        return {};
    const data = JSON.parse(Buffer.from(identifier, 'base64').toString('utf8'));
    return fields.reduce((o, e, i) => {
        let op = '$gt';
        if (sort[e] < 0)
            op = '$lt';
        o[e] = { [op]: data[i] };
        return o;
    }, {});
}
exports.getCursorQuery = getCursorQuery;
function getNextCursor(dataSet, cursor) {
    const nCursor = {
        sort: cursor.sort,
        limit: cursor.limit,
        next: undefined,
        current: cursor.next || cursor.current,
    };
    if (!dataSet.length)
        return nCursor;
    const last = dataSet[dataSet.length - 1];
    if (!last)
        return nCursor;
    const fields = Object.keys(cursor.sort).sort();
    if (fields.length > 1)
        throw new Error('Cursors with multiple sort fields are not supported yet');
    nCursor.next = Buffer.from(JSON.stringify(fields.map(k => {
        const segments = k.split('.');
        let target = last;
        for (const k of segments) {
            if (!target || typeof target !== 'object')
                return target;
            target = target[k];
        }
        if (target instanceof Date)
            return target.toISOString();
        return target;
    }))).toString('base64');
    return nCursor;
}
exports.getNextCursor = getNextCursor;
class RequestError extends Error {
    constructor(message, code = '500') {
        super(message);
        this.code = '' + code;
    }
}
exports.RequestError = RequestError;
function isRequest(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === undefined)
        && (target.action)
        && (typeof target.action === 'string')
        && (target.body === undefined || (target.body && (typeof target.body === 'object')))
        && (target.resources)
        && (typeof target.resources === 'object')
        && (target.resources.length !== undefined)
        && (target.origin === undefined || (target.origin
            && typeof target.origin === 'object'
            && typeof target.origin.id === 'string'
            && typeof target.origin.type === 'string'))
        && (target.timeout === undefined || (['string', 'number'].includes(typeof target.timeout)
            && !isNaN(+target.timeout))));
}
exports.isRequest = isRequest;
