// 55tec_integration_lib/util/auth.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const index_1 = require("../model/index");
const index_2 = require("../model/metadata/action/index");
const entity_1 = require("../model/metadata/action/entity");
const operation_1 = require("../model/metadata/action/operation");
class Auth {
    constructor(integrationClient) {
        this.integrationClient = integrationClient;
    }
    static makeToken(clientId) {
        return jsonwebtoken_1.default.sign({ client_id: clientId }, process.env.JWT_SECRET || '', { expiresIn: 86400000 });
    }
    async getClient(search) {
        const { credentials, clientId } = search;
        const response = await this.integrationClient.send(index_1.Integration.ADMIN, index_2.newAction(entity_1.Entity.Client, operation_1.Operation.FIND), {
            data: {
                clientId
            },
            credentials
        });
        return response.data;
    }
    async getAuthToken(search) {
        const { clientId } = await this.getClient(search);
        return Auth.makeToken(clientId);
    }
}
exports.default = Auth;
