// 55tec_integration_lib/util/promise.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IdentifiablePromise = exports.Timeout = exports.NamedError = void 0;
class NamedError extends Error {
    constructor(message, name) {
        super(message);
        this.name = name;
        Object.setPrototypeOf(this, NamedError.prototype);
    }
    toJSON() {
        return {
            message: this.message,
            name: this.name,
        };
    }
}
exports.NamedError = NamedError;
class Timeout {
    constructor(resolution, time) {
        this.time = time;
        this.createdAt = new Date();
        if (time !== Infinity)
            this.id = setTimeout(() => {
                this.id = undefined;
                resolution();
            }, time);
    }
    stop() {
        if (this.id)
            clearTimeout(this.id);
    }
    timeLeft() {
        return Math.max(0, this.time - ((new Date()).getTime() - this.createdAt.getTime()));
    }
}
exports.Timeout = Timeout;
class IdentifiablePromise {
    constructor(id, context, registry) {
        this.id = id;
        this.done = false;
        this.registry = registry;
        this.context = context;
        if (this.registry)
            this.registry[this.id] = this;
        this.resolve = (res) => IdentifiablePromise.DEFAULT_RESOLUTION(this);
        this.reject = (err) => IdentifiablePromise.DEFAULT_RESOLUTION(this);
        this.promise = new Promise((resolve, reject) => {
            this.resolve = (res) => IdentifiablePromise.DEFAULT_RESOLUTION(this) && resolve(res);
            this.reject = (err) => IdentifiablePromise.DEFAULT_RESOLUTION(this) && reject(err);
        });
    }
    static DEFAULT_RESOLUTION(req) {
        req.done = true;
        if (req.registry)
            delete req.registry[req.id];
        if (req.timeout)
            req.timeout.stop();
        return true;
    }
    then(onFulfilled, onRejected) {
        return this.promise = this.promise.then(onFulfilled, onRejected);
    }
    catch(handler) {
        return this.promise = this.promise.catch(handler);
    }
    finally(handler) {
        return this.promise = this.promise.finally(handler);
    }
    isDone() { return this.done; }
    setTimeout(time) {
        this.timeout = new Timeout(() => {
            if (this.done)
                return;
            this.reject(new NamedError(`The request "${this.id}" timed out after ${time}ms`, 'message_timeout'));
        }, time);
    }
    untilTimeout() {
        if (!this.timeout)
            return Infinity;
        return this.timeout.timeLeft();
    }
    get [Symbol.toStringTag]() {
        return 'IdentifiablePromise';
    }
}
exports.IdentifiablePromise = IdentifiablePromise;
