// 55tec_integration_lib/service/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Context = void 0;
require("reflect-metadata");
const inversify_1 = require("inversify");
const request_promise_1 = __importDefault(require("request-promise"));
const url_1 = require("url");
const querystring_1 = __importDefault(require("querystring"));
const request_1 = require("../model/protocol/integrator/request");
const response_1 = require("../model/protocol/integrator/response");
const credential_1 = require("../model/protocol/integrator/credential");
const protocol_1 = require("../model/protocol");
let Service = class Service {
    async request(url, method, body, headers = {}, secure = true) {
        if (body && method === "get") {
            let parsed = new url_1.URL(url);
            body = {
                ...querystring_1.default.parse(parsed.search),
                ...body
            };
            url = `${parsed.origin}${parsed.pathname}?${querystring_1.default.stringify(body)}${parsed.hash}`;
        }
        headers = {
            'Content-Type': 'application/json',
            ...headers
        };
        let result = await request_promise_1.default({
            "rejectUnauthorized": secure,
            "url": url,
            "method": method,
            "body": method !== "get" ? JSON.stringify(body) : undefined,
            "headers": headers,
            "simple": false,
            "resolveWithFullResponse": true
        });
        if (("" + result.statusCode)[0] !== "2")
            throw new request_1.RequestError(result.body, result.statusCode);
        if (!result.body)
            return {};
        let data = result.body;
        if (typeof data === "string")
            data = JSON.parse(result.body.trim() || "{}");
        return data;
    }
    async save(ctx) {
        throw new response_1.ResponseError(`Not implemented yet`, protocol_1.StatusCode.SERVICE_UNAVAILABLE);
    }
    async find(ctx) {
        throw new response_1.ResponseError(`Not implemented yet`, protocol_1.StatusCode.SERVICE_UNAVAILABLE);
    }
    async list(ctx) {
        throw new response_1.ResponseError(`Not implemented yet`, protocol_1.StatusCode.SERVICE_UNAVAILABLE);
    }
    async delete(ctx) {
        throw new response_1.ResponseError(`Not implemented yet`, protocol_1.StatusCode.SERVICE_UNAVAILABLE);
    }
    getWWWURL(entity, id, ctx) {
        throw new response_1.ResponseError(`Not implemented yet`, protocol_1.StatusCode.SERVICE_UNAVAILABLE);
    }
};
Service.DEFAULT_PAGE_SIZE = 100;
Service = __decorate([
    inversify_1.injectable()
], Service);
exports.default = Service;
class Context {
    constructor(req) {
        this.auth = {};
        this.id = req.id;
        if (!req.body)
            throw new request_1.RequestError(`Missing request body: ${JSON.stringify({ ...req, "credentials": undefined })}`, "400");
        this.action = req.action;
        let { credentials = [], projection = [], data = {}, pagination = {
            "page": 0,
            "size": Service.DEFAULT_PAGE_SIZE
        }, cursor, options = {} } = req.body;
        this.credentials = credentials;
        for (let c of credentials) { // @ts-ignore
            this.auth[c.type] = c.value;
        }
        this.payload = data;
        this.opts = options;
        // if (!cursor) cursor = { limit: pagination.size, current: pagination.page };
        this.params = { projection, pagination, cursor };
        this.resourceMap = (req.resources || []).reduce((map, e) => {
            map[e.type] = e;
            return map;
        }, {});
        this.origin = req.origin;
        this.timeout = req.timeout;
    }
    getAuthHeader() {
        // if (this.auth[Type.KEY]) return {'Authorization': `Bearer ${this.auth[Type.KEY]}`};
        if (this.auth[credential_1.Type.TOKEN])
            return { 'Authorization': `Bearer ${this.auth[credential_1.Type.TOKEN]}` };
        if (this.auth[credential_1.Type.BASIC]) {
            let { id, secret } = this.auth[credential_1.Type.BASIC];
            return { 'Authorization': `Basic ${Buffer.from(`${id}:${secret}`).toString("base64")}` };
        }
        return {};
    }
    hasResource(type) {
        return !!this.resourceMap[type];
    }
}
exports.Context = Context;
