// 55tec_integration_lib/model/protocol/server/message/registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.isResponse = exports.isRequest = void 0;
const uuid_1 = require("uuid");
const __1 = require("../");
const resource_1 = require("../resource");
const __2 = require("../../");
const index_1 = require("../../index");
function isRequest(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === 'server')
        && (target.type === __1.MessageType.REGISTRY)
        && (target.body)
        && (typeof target.body === 'object')
        && (Object.keys(target.body).every(k => Object.values(resource_1.Method).includes(k))));
}
exports.isRequest = isRequest;
function isResponse(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === 'server')
        && (target.type === __1.MessageType.REGISTRY)
        && (!target.error || (typeof target.error === 'string')));
}
exports.isResponse = isResponse;
class Client extends __2.Client {
    constructor(channel) {
        super(channel);
    }
    async build(data) {
        return {
            id: uuid_1.v4(),
            channel: 'server',
            type: __1.MessageType.REGISTRY,
            body: Object.keys(data.resources).reduce((o, method) => {
                let registry = data.resources[method];
                o[method] = Object.keys(registry);
                return o;
            }, {})
        };
    }
    async send(router) {
        return super.send(router);
    }
    isRequest(target) { return isRequest(target); }
    isResponse(target) { return isResponse(target); }
    implementsProtocol(target) {
        return this.isRequest(target) || this.isResponse(target);
    }
    async handleUnregistredMessage(message) { }
    parseResolution(data) {
        if (!this.isResponse(data))
            throw new index_1.ResolutionError('response', data);
        if (data.error)
            throw new Error(data.error);
        return {};
    }
}
exports.Client = Client;
