// 55tec_integration_lib/model/protocol/server/message/route.js
"use strict";
/**
 * @TODO refactor names
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.isResponse = exports.isRequest = void 0;
const __1 = require("../");
const resource_1 = require("../resource");
const __2 = require("../../");
function isRequest(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === 'server')
        && (target.type === __1.MessageType.ROUTE)
        && (target.body && (typeof target.body === 'object'))
        && (typeof target.body.urn === 'string')
        && (Object.values(resource_1.Method).includes(target.body.method))
        && (target.body.query && (typeof target.body.query === 'object'))
        && (target.body.body && (typeof target.body.body === 'object'))
        && (target.body.params && (typeof target.body.params === 'object'))
        && (target.body.headers && (typeof target.body.headers === 'object')));
}
exports.isRequest = isRequest;
function isResponse(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === 'server')
        && (target.type === __1.MessageType.ROUTE)
        && (!target.error || (typeof target.error === 'string')));
}
exports.isResponse = isResponse;
class Client extends __2.Client {
    constructor(channel) {
        super(channel);
        this.REQUEST_HANDLER_TIMEOUT = 20000;
        this.routers = [];
    }
    add(router) {
        this.routers.push(router);
    }
    async build(id, data, code = '200', headers) {
        let res = {
            id,
            channel: 'server',
            code,
            headers,
            type: __1.MessageType.ROUTE,
        };
        if (data instanceof resource_1.HTTPError) {
            res.error = data.message;
            res.code = data.code;
            res.name = data.name;
        }
        else
            res.body = data;
        return res;
    }
    async send(id, data, code, headers, name) {
        return super.send(id, data, code, headers, name);
    }
    isRequest(target) { return isRequest(target); }
    isResponse(target) { return isResponse(target); }
    implementsProtocol(target) {
        return this.isRequest(target) || this.isResponse(target);
    }
    async handleUnregistredMessage(message) {
        if (!this.isRequest(message))
            return;
        let id = message.id;
        let { urn, method } = message.body;
        let router = this.routers.find(router => !!router.getHandler(method, urn));
        if (!router)
            return this.send(id, new resource_1.HTTPError('Not found', '404'));
        let handler = router.getHandler(method, urn);
        try {
            let literal = await this.process(message, handler);
            return this.send(id, literal.body, literal.code, literal.headers);
        }
        catch (err) {
            if (!(err instanceof resource_1.HTTPError))
                err = new resource_1.HTTPError(err.message);
            return this.send(id, err, err.code, {}, err.name);
        }
    }
    async process(req, handler) {
        let reqWrapper = new resource_1.RequestImpl(req.body);
        let resWrapper = new resource_1.ResponseImpl(req.id, req.body);
        handler(reqWrapper, resWrapper);
        if (this.REQUEST_HANDLER_TIMEOUT > 0)
            resWrapper.setTimeout(this.REQUEST_HANDLER_TIMEOUT);
        return resWrapper;
    }
    parseResolution(data) {
        // no resolution is needed if no requests are sent
    }
}
exports.Client = Client;
