// 55tec_integration_lib/model/metadata/action/entity.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Entity = exports._Entity = void 0;
const pluralize_1 = __importDefault(require("pluralize"));
var _Entity;
(function (_Entity) {
    _Entity["CUSTOMER"] = "customers";
    _Entity["INCIDENT"] = "incidents";
    _Entity["USER"] = "users";
    _Entity["PIPELINE"] = "pipelines";
    _Entity["CALL"] = "calls";
    _Entity["CHAT"] = "chats";
    _Entity["MESSAGE"] = "messages";
    _Entity["SALE"] = "sales";
    _Entity["FUNCTION"] = "functions";
    _Entity["BOT"] = "bots";
    _Entity["TEMPLATE"] = "templates";
    _Entity["REQUEST"] = "requests";
})(_Entity = exports._Entity || (exports._Entity = {}));
exports.Entity = new Proxy({ ..._Entity }, {
    get: (target, p, receiver) => {
        if (typeof p === 'symbol')
            return undefined;
        if (typeof p === 'number')
            return undefined;
        return target[p] || p.split('/').map(p => pluralize_1.default(('' + p).toLowerCase().replace(/_/g, '-'))).join('/');
    }
});
