// 55tec_integration_lib/model/metadata/action/operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Operation = exports.Browser = exports.Meta = exports.Auth = exports.CRUD = void 0;
var CRUD;
(function (CRUD) {
    CRUD["CREATE"] = "create";
    CRUD["FIND"] = "find";
    CRUD["LIST"] = "list";
    CRUD["UPDATE"] = "update";
    CRUD["DELETE"] = "delete";
})(CRUD = exports.CRUD || (exports.CRUD = {}));
var Auth;
(function (Auth) {
    Auth["REFRESH_TOKEN"] = "refresh-token";
    Auth["GET_TOKEN"] = "get-token";
    Auth["GET_OAUTH_URL"] = "get-oauth-url";
    Auth["GET_DEVICE_CODE"] = "get-device-code";
    Auth["GET_SESSION"] = "get-session";
})(Auth = exports.Auth || (exports.Auth = {}));
var Meta;
(function (Meta) {
    Meta["GET_METADATA"] = "get-metadata";
})(Meta = exports.Meta || (exports.Meta = {}));
var Browser;
(function (Browser) {
    Browser["SETUP"] = "setup";
    Browser["SHOW"] = "show";
    Browser["RESIZE"] = "resize";
})(Browser = exports.Browser || (exports.Browser = {}));
exports.Operation = new Proxy({ ...CRUD, ...Auth, ...Meta, ...Browser }, {
    get: (target, p, receiver) => {
        if (typeof p === 'symbol')
            return undefined;
        if (typeof p === 'number')
            return undefined;
        return target[p] || ('' + p).toLowerCase().replace(/_/g, '-');
    }
});
