// 55tec_integration_lib/model/protocol/server/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Router = exports.Client = exports.isResponse = exports.isRequest = exports.MessageType = void 0;
const protocol_1 = require("../../protocol");
const registry_1 = require("./message/registry");
const route_1 = require("./message/route");
var MessageType;
(function (MessageType) {
    MessageType["REGISTRY"] = "registry";
    MessageType["ROUTE"] = "route";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
function isRequest(target) {
    return registry_1.isRequest(target) || route_1.isRequest(target);
}
exports.isRequest = isRequest;
function isResponse(target) {
    return registry_1.isResponse(target) || route_1.isResponse(target);
}
exports.isResponse = isResponse;
/**
 * @TODO refactor to make it respect liskov principle
 */
class Client extends protocol_1.Client {
    constructor(channel) {
        super(channel);
        this.registryClient = new registry_1.Client(channel);
        this.routeClient = new route_1.Client(channel);
    }
    async use(router) {
        await new Promise(resolve => setTimeout(resolve, 3000));
        await this.registryClient.send(router);
        this.routeClient.add(router);
    }
    async build() {
        throw new Error('Method "build" cannot be called externally. Use "use" method instead');
    }
    async send() {
        throw new Error('Method "send" cannot be called externally. Use "use" method instead');
    }
    isRequest(target) { return isRequest(target); }
    isResponse(target) { return isResponse(target); }
    implementsProtocol(target) {
        return this.isRequest(target) || this.isResponse(target);
    }
    async handleUnregistredMessage(message) { }
    parseResolution(data) { }
    setTimeout(timeout) {
        this.routeClient.REQUEST_HANDLER_TIMEOUT = timeout;
    }
}
exports.Client = Client;
var resource_1 = require("./resource");
Object.defineProperty(exports, "Router", { enumerable: true, get: function () { return resource_1.Router; } });
