// 55tec_integration_lib/model/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Integration = void 0;
var Integration;
(function (Integration) {
    Integration["ADOBE_HENDRIX"] = "adobe-hendrix";
    Integration["ADMIN"] = "admin";
    Integration["AUTH"] = "auth";
    Integration["AZURE"] = "azure";
    Integration["CHAT"] = "chat";
    Integration["CHATBOT"] = "chatbot";
    Integration["CHAT_SDK"] = "chat-sdk";
    Integration["DATA"] = "data";
    Integration["DATA_AGGREGATE"] = "data-aggregate";
    Integration["DIALER"] = "dialer";
    Integration["DEBUG"] = "debug";
    Integration["EMAIL"] = "email";
    Integration["FLOW"] = "flow";
    Integration["FRESHDESK"] = "freshdesk";
    Integration["FRESHSALES"] = "freshsales";
    Integration["HUBSPOT"] = "hubspot";
    Integration["MESSAGING_BRIDGE"] = "messaging-bridge";
    Integration["MOVIDESK"] = "movidesk";
    Integration["PIPEDRIVE"] = "pipedrive";
    Integration["POWERBI"] = "powerbi";
    Integration["SALESFORCE"] = "salesforce";
    Integration["SALESFORCE_EMBEDDED"] = "salesforce-embedded";
    Integration["SAP"] = "sap";
    Integration["SMS"] = "sms";
    Integration["SUPPORT"] = "support";
    Integration["TELEGRAM"] = "telegram";
    Integration["TWILIO"] = "twilio";
    Integration["WHATSAPP"] = "whatsapp";
    Integration["ZENDESK"] = "zendesk";
    Integration["ZENDESK_SELL"] = "zendesk-sell";
    Integration["INSTAGRAM"] = "instagram";
    Integration["ZOHO"] = "zoho";
    Integration["BILLING"] = "billing";
    Integration["GUPSHUP"] = "gupshup";
    Integration["BROKER_GUPSHUP"] = "broker-gupshup";
    Integration["MANDRILLAPP"] = "mandrillapp";
    Integration["AXIOS"] = "axios";
})(Integration = exports.Integration || (exports.Integration = {}));
