// 55tec_integration_lib/model/protocol/integrator/response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseError = exports.isResponse = void 0;
const __1 = require("../");
function isResponse(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === undefined)
        && (target.response)
        && (typeof target.response === 'object')
        && (target.response.code)
        && (Object.values(__1.StatusCode).includes(('' + target.response.code))));
}
exports.isResponse = isResponse;
class ResponseError extends Error {
    constructor(message, code, name) {
        super(message);
        if (code == __1.StatusCode.OK)
            throw new Error(`Invalid error code: ${code}`);
        this.message = message;
        this.code = code;
        this.name = name || '';
    }
    toJSON() {
        return {
            message: this.message,
            code: this.code,
            name: this.name,
        };
    }
    toString() {
        return `${this.name} - ${this.code} - ${this.message}`;
    }
}
exports.ResponseError = ResponseError;
