// 55tec_integration_lib/model/protocol/file/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.isResponse = exports.isRequest = exports.Operation = void 0;
const fs_1 = __importDefault(require("fs"));
const querystring_1 = __importDefault(require("querystring"));
const request_promise_1 = __importDefault(require("request-promise"));
const config_pbx_1 = __importDefault(require("config-pbx"));
const auth_1 = __importDefault(require("../../../util/auth"));
const protocol_1 = require("../../protocol");
var Operation;
(function (Operation) {
    Operation["DOWNLOAD"] = "download";
    Operation["UPLOAD"] = "upload";
})(Operation = exports.Operation || (exports.Operation = {}));
function isRequest(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === 'file')
        && (target.body)
        && (typeof target.body === 'object')
        && (typeof target.body.name === 'string')
        && (typeof target.body.mimeType === 'string')
        && (typeof target.body.repository === 'string')
        && (typeof target.body.operation === 'string')
        && (typeof target.body.credentials === 'object')
        && (!target.body.uri || typeof target.body.uri === 'string')
        && (!target.body.data || typeof target.body.data === 'string'));
}
exports.isRequest = isRequest;
function isResponse(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === 'file')
        && (!target.error || (typeof target.error === 'string'))
        && (!target.body || (typeof target.body === 'object'
            && typeof target.body.name === 'string'
            && typeof target.body.mimeType === 'string'
            && typeof target.body.repository === 'string'
            && typeof target.body.createdAt === 'string'
            && typeof target.body.ttl === 'number'
            && (!target.body.uri || (typeof target.body.uri === 'string'))
            && (!target.body.data || (typeof target.body.data === 'string')))));
}
exports.isResponse = isResponse;
/**
 * @TODO refactor to make it respect liskov principle
 */
class Client extends protocol_1.Client {
    constructor(channel, authUtils) {
        super(channel);
        this.authUtils = authUtils;
    }
    async downloadByStream(credentials, repository, filename, writable) {
        const [token, sip] = await this.getTokenAndSip(credentials);
        const filesBase = await this.getFileBase(sip);
        const uri = `${filesBase}/${repository}/${filename}`;
        request_promise_1.default.get(uri, {
            headers: {
                'authorization': `Bearer ${token}`
            },
        }).pipe(writable);
    }
    async download(credentials, repository, filename, destination) {
        const file = fs_1.default.createWriteStream(destination);
        await this.downloadByStream(credentials, repository, filename, file);
        return new Promise((resolve, reject) => {
            file.on('finish', () => resolve());
            file.on('error', (err) => reject(err));
        });
    }
    async uploadByStream(credentials, descriptorOrRepository, readable) {
        const [token, sip] = await this.getTokenAndSip(credentials);
        let repository = descriptorOrRepository;
        let query = '';
        if (typeof descriptorOrRepository !== 'string') {
            repository = descriptorOrRepository.repository;
            query = '?' + querystring_1.default.stringify(descriptorOrRepository);
        }
        const filesBase = await this.getFileBase(sip);
        let uri = `${filesBase}/${repository}${query}`;
        return new Promise((resolve, reject) => {
            request_promise_1.default.post(uri, {
                headers: {
                    'authorization': `Bearer ${token}`,
                },
                formData: {
                    file: readable
                },
            }, (err, response) => {
                if (err)
                    reject(err);
                resolve(JSON.parse(response.body));
            });
        });
    }
    async upload(credentials, file) {
        const [token, sip] = await this.getTokenAndSip(credentials);
        const filesBase = await this.getFileBase(sip);
        const uri = `${filesBase}/${file.repository}`;
        return new Promise((resolve, reject) => {
            request_promise_1.default.post(uri, {
                headers: {
                    'authorization': `Bearer ${token}`,
                    'content-type': 'application/json'
                },
                body: JSON.stringify({ file }),
            }, (err, response) => {
                if (err)
                    reject(err);
                resolve(JSON.parse(response.body));
            });
        });
    }
    async build() {
        throw new Error('Method "build" cannot be called externally. Use "upload", "download" or "download" method instead');
    }
    async send() {
        throw new Error('Method "send" cannot be called externally. Use "upload", "download" or "download" method instead');
    }
    isRequest(target) { return isRequest(target); }
    isResponse(target) { return isResponse(target); }
    implementsProtocol(target) {
        return this.isRequest(target) || this.isResponse(target);
    }
    async handleUnregistredMessage(message) { }
    parseResolution(data) { }
    async getFileBase(sipServer) {
        let applications = config_pbx_1.default.get('application-config', '' + process.env.NODE_ENV, { sipServer });
        if (applications && typeof applications.then === 'function')
            applications = await applications;
        const application = applications.find((e) => e.identifier === '55pbx');
        if (!application)
            throw new Error(`Application settings not found`);
        const sip = application.settings.sip.find((sip) => sip.sip_server === sipServer);
        if (!sip)
            throw new Error(`SIP settings not found: ${sipServer}`);
        return sip.files_base;
    }
    async getTokenAndSip(credentials) {
        let search;
        if (typeof credentials === 'string')
            search = { clientId: credentials };
        else
            search = { credentials };
        const { clientId, sipServer } = await this.authUtils.getClient(search);
        const token = auth_1.default.makeToken(clientId);
        return [token, sipServer];
    }
}
exports.Client = Client;
