// 55tec_integration_lib/model/protocol/browser/func.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlertType = exports.ParticipantType = exports.ContactType = exports.MessageType = void 0;
var MessageType;
(function (MessageType) {
    MessageType["TEXT"] = "text";
    MessageType["ATTACHMENT"] = "attachment";
    MessageType["FORM"] = "form";
    MessageType["INPUT"] = "input";
    MessageType["SYSTEM"] = "system";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
var ContactType;
(function (ContactType) {
    ContactType["BRANCH"] = "branch";
    ContactType["PHONE"] = "phone";
    ContactType["CODE"] = "code";
})(ContactType = exports.ContactType || (exports.ContactType = {}));
var ParticipantType;
(function (ParticipantType) {
    ParticipantType["USER"] = "user";
    ParticipantType["CUSTOMER"] = "customer";
    ParticipantType["MENU"] = "menu";
    ParticipantType["QUEUE"] = "queue";
    ParticipantType["ANONYMOUS"] = "anonymous";
    ParticipantType["SYSTEM"] = "system";
})(ParticipantType = exports.ParticipantType || (exports.ParticipantType = {}));
var AlertType;
(function (AlertType) {
    AlertType["ERROR"] = "error";
    AlertType["WARNING"] = "warning";
    AlertType["INFO"] = "info";
})(AlertType = exports.AlertType || (exports.AlertType = {}));
