// 55tec_integration_lib/model/protocol/browser/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.generateBrowserId = exports.isResponse = exports.isRequest = exports.TargetType = exports.CacheMissError = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const uuid_1 = require("uuid");
const __1 = require("../");
const protocol_1 = require("../../protocol");
class CacheMissError extends Error {
    constructor(message) {
        super(`The function was not found on integrator cache - please send the full payload - ${message}`);
        Object.setPrototypeOf(this, CacheMissError.prototype);
    }
}
exports.CacheMissError = CacheMissError;
var TargetType;
(function (TargetType) {
    TargetType["ADMIN"] = "admin";
    TargetType["WEBPHONE"] = "webphone";
    TargetType["ALL"] = "all";
})(TargetType = exports.TargetType || (exports.TargetType = {}));
function isRequest(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === 'browser')
        && (typeof target.browserId === 'string')
        && (target.body)
        && (typeof target.body === 'object')
        && (target.body.params)
        && (typeof target.body.params === 'object')
        && (target.body.params.length !== undefined)
        && (!target.body.setup || typeof target.body.setup === 'string')
        && (!target.body.func || typeof target.body.func === 'string')
        && (!target.body.name || typeof target.body.name === 'string'));
}
exports.isRequest = isRequest;
function isResponse(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === 'browser')
        && (!target.error || (typeof target.error === 'string')));
}
exports.isResponse = isResponse;
function generateBrowserId(clientId, branchNumber) {
    return `${clientId}-${branchNumber}`;
}
exports.generateBrowserId = generateBrowserId;
class Client extends protocol_1.Client {
    constructor(channel, funcDir) {
        super(channel);
        this.funcDir = funcDir;
        this.funcCache = {};
    }
    async load(funcName) {
        let func = await (new Promise((resolve, reject) => fs_1.default.readFile(path_1.default.join(this.funcDir, `./${funcName}.js`), (err, data) => err ? reject(err) : resolve(data.toString('ascii')))));
        return this.funcCache[funcName] = func;
    }
    async build(browserId, funcName, data) {
        const id = uuid_1.v4();
        const body = {
            name: funcName,
            params: data,
        };
        if (!this.funcCache[funcName])
            body.func = await this.load(funcName);
        if (funcName !== 'setup' && !this.funcCache['setup'])
            body.setup = await this.load('setup');
        return {
            id,
            browserId,
            channel: 'browser',
            body,
        };
    }
    async send(browserId, funcName, data, config = { target: TargetType.WEBPHONE }) {
        try {
            let result = await super.send(browserId, funcName, data, config);
            return result;
        }
        catch (err) {
            if (err instanceof CacheMissError) {
                delete this.funcCache[funcName];
                delete this.funcCache['setup'];
                return super.send(browserId, funcName, data, config);
            }
            throw err;
        }
    }
    isRequest(target) { return isRequest(target); }
    isResponse(target) { return isResponse(target); }
    implementsProtocol(target) {
        return this.isRequest(target) || this.isResponse(target);
    }
    async handleUnregistredMessage(message) { }
    parseResolution(data) {
        if (!this.isResponse(data))
            throw new __1.ResolutionError('response', data);
        if (data.error) {
            if (data.error.toLowerCase().includes('cache miss'))
                throw new CacheMissError(data.error);
            throw new Error(data.error);
        }
        return data.body;
    }
}
exports.Client = Client;
