// 55tec_integration_lib/model/protocol/integration/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.isResponse = exports.isRequest = void 0;
const uuid_1 = require("uuid");
const __1 = require("../");
const response_1 = require("../integrator/response");
const __2 = require("../");
const __3 = require("../");
function isRequest(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === 'integration')
        && (target.action)
        && (typeof target.action === 'string')
        && (target.body === undefined || (target.body && (typeof target.body === 'object'))));
}
exports.isRequest = isRequest;
function isResponse(target) {
    return (!!target
        && (typeof target === 'object')
        && (typeof target.id === 'string')
        && (target.channel === 'integration')
        && (target.response)
        && (typeof target.response === 'object')
        && (Object.values(__2.StatusCode).includes(('' + target.response.code)))
        && (!target.response.error
            || ((typeof target.response.error === 'string')
                || (typeof target.response.error === 'object'
                    && target.response.error.message
                    && typeof target.response.error.message === 'string'))));
}
exports.isResponse = isResponse;
/**
 * @TODO implements resource discovery to avoid having to make all actions hardcoded
 */
class Client extends __1.Client {
    async send(integration, action, body) {
        return super.send(integration, action, body);
    }
    isRequest(target) { return isRequest(target); }
    isResponse(target) { return isResponse(target); }
    implementsProtocol(target) {
        return this.isRequest(target) || this.isResponse(target);
    }
    async handleUnregistredMessage(message) { }
    async build(integration, action, body) {
        return {
            id: uuid_1.v4(),
            channel: 'integration',
            action,
            integration,
            resources: [],
            body,
        };
    }
    parseResolution(data) {
        if (!this.isResponse(data))
            throw new __3.ResolutionError('response', data);
        if (data.response.error)
            throw new response_1.ResponseError(typeof data.response.error === 'string' ? data.response.error : data.response.error.message, data.response.code, data.response.error && typeof data.response.error === 'object' ? data.response.error.name : "");
        return data.response.body;
    }
}
exports.Client = Client;
